src/recipes.jl

```diff
@@ -378,109 +378,256 @@ end
-    counts = zeros(length(edges)-1)
+@recipe function f(::Type{Val{:stepbins}}, x, y, z)
+    edge, weights = x, y
+
+    axis = d[:subplot][Plots.isvertical(d) ? :xaxis : :yaxis]
 
-    # add a weighted count
-    for (i,vi) in enumerate(v)
-        idx = bucket_index(vi, edges)
-        counts[idx] += (weights == nothing ? 1.0 : weights[i])
+    xpts, ypts = _stepbins_path(edge, weights)
+    if !Plots.isvertical(d)
+        xpts, ypts = ypts, xpts
     end
 
-    counts = isapprox(extrema(diff(edges))...) ? counts : counts ./ diff(edges) # for uneven bins, normalize to area.
+    # create a secondary series for the markers
+    if d[:markershape] != :none
+        @series begin
+            seriestype := :scatter
+            x := Plots._bin_centers(edge)
+            y := weights
+            fillrange := nothing
+            label := ""
+            primary := false
+            ()
+        end
+        markershape := :none
+        xerror := :none
+        yerror := :none
+    end
 
-    # normalize by bar area?
-    norm_denom = normed ? sum(diff(edges) .* counts) : 1.0
-    if norm_denom == 0
-        norm_denom = 1.0
+    x := xpts
+    y := ypts
+    seriestype := :path
+    ylims --> [0, 1.1 * maximum(weights)]
+    ()
+end
+Plots.@deps stepbins path
+
+
+function _auto_binning_nbins{N}(vs::NTuple{N,AbstractVector}, dim::Integer; mode::Symbol = :auto)
+    _cl(x) = max(ceil(Int, x), 1)
+    _iqr(v) = quantile(v, 0.75) - quantile(v, 0.25)
+    _span(v) = maximum(v) - minimum(v)
+
+    n_samples = length(linearindices(first(vs)))
+    # Estimator for number of samples in one row/column of bins along each axis:
+    n = max(1, n_samples^(1/N))
+
+    v = vs[dim]
+
+    if mode == :sqrt  # Square-root choice
+        _cl(sqrt(n))
+    elseif mode == :sturges || mode ==:auto  # Sturges' formula
+        _cl(log2(n)) + 1
+    elseif mode == :rice  # Rice Rule
+        _cl(2 * n^(1/3))
+    elseif mode == :scott  # Scott's normal reference rule
+        _cl(_span(v) / (3.5 * std(v) / n^(1/3)))
+    elseif mode == :fd  # Freedman–Diaconis rule
+        _cl(_span(v) / (2 * _iqr(v) / n^(1/3)))
+    else
+        error("Unknown auto-binning mode $mode")
     end
+end
+
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::Integer) = StatsBase.histrange(vs[dim], binning, :left)
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::Symbol) = _hist_edge(vs, dim, _auto_binning_nbins(vs, dim, mode = binning))
+_hist_edge{N}(vs::NTuple{N,AbstractVector}, dim::Integer, binning::AbstractVector) = binning
 
-    edges, counts ./ norm_denom
+_hist_edges{N}(vs::NTuple{N,AbstractVector}, binning::NTuple{N}) =
+    map(dim -> _hist_edge(vs, dim, binning[dim]), (1:N...))
+
+_hist_edges{N}(vs::NTuple{N,AbstractVector}, binning::Union{Integer, Symbol, AbstractVector}) =
+    map(dim -> _hist_edge(vs, dim, binning), (1:N...))
+
+_hist_norm_mode(mode::Symbol) = mode
+_hist_norm_mode(mode::Bool) = mode ? :norm : :none
```

Comment:
> Our difference is due to different views of what a histogram is. I think of it as primarily a visualization of the underlying statistical distribution. You may think of it as primarily a bin-counting technique

It can be either, of course. However, I think it's pretty undisputed that for "normal" histograms (homogenous edge(s)) the vast majority of users will expect to see counts on the y-axis. Are there any major software packages that automatically normalize homogenously binned histograms by default? Naturally, for inhomogenous bins, plotting counts will often not make sense  (except for cross-checks, etc.). One will usually want to plot a PDF estimate or density - but Plots.jl has no way to guess which of those the user wants. And if the user doesn't know what he wants, he can't interpret the results, anyhow. My argument is not that hists with inhomogenous bins shouldn't be normalized, but that it shouldn't be done automatically.

> Your best bet at convincing me is to find papers using uneven bins without density normalization

In my opinion, this has nothing to do with statistical theory (like I said, of course the user should normalize in some fashion in these cases). It's a question of software design. In general, the behavior of Plots (or any other package) should be predictable for the users - automatically plotting either counts or density goes against that.

There's also another problem: How would we reliably detect inhomogenous edges? The edge definition might very well come from a source with limited numerical precision, e.g. a text file. Now where do we set the numerical threshold for "bins are of equal width"? If a user supplies bin edges directly, instead of using a range, the choice of counts vs. density would really become random, depending on the precision of his values.